### ISSUE TITLE
Bug: clicking on the label + checkbox, NVDA is announcing inverted values of checkbox.

### ISSUE THREAD

**Issue by franzitaz** [nvaccess/nvda#15854]
### Steps to reproduce:
1. Open NVDA, version 2023.3
2. Open this link for test and screenshots: [NVDA testing input checkbox + label](https://codepen.io/franzitaz/pen/bGzxwqE)

### Actual behavior:

When we click on the **label**, NVDA is announcing inverted values of checkbox.
**Ps:** if we add a **::before pseudo-element** to label, NVDA announces the expected values of checkbox.

### Expected behavior:

After clicking on the **label**, the expected is NVDA announces the correct value of checkbox when is checked or not.

#### NVDA installed/portable/running from source:
Installed

---
#### NVDA version:
2023.3

---
#### Windows version:
Version	10.0.19045 Compilation 19045

---
#### Does the issue still occur after restarting your computer?
Yes

---
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
No, I didn't find on NVDA website a previous version to test.

---
#### Screenshots:
![example 1](https://github.com/nvaccess/nvda/assets/11142312/c0bf8335-667f-4c5a-aa57-9aa7d535e9c9)
![example 2-before pesudo-element](https://github.com/nvaccess/nvda/assets/11142312/25e82c5e-1871-4534-819d-4ffb3510fe9a)

**Comment by lukaszgo1** [comment]
Please mention in which web browsers (mentioning exact version numbers)  this can be reproduced.

**Comment by Adriani90** [comment]
I think this is only reproducible in Chrome and Edge, because in Firefox NVDA does not announce the change of the state of the checkboxes at all in focus mode or when using the mouse.
@jcsteh what is the reason for Firefox not anouncing the state change at all?

Using the "before" or "after" atribute seems to lead to failure of success criterion 1.3.1 in WCAG:
https://www.w3.org/TR/WCAG20-TECHS/F87.html

I think supporting this correctly needs changes on both, screenreader and browsers side.
I personally would like to have consistent support for this because it could help in dynamic situations such as video conferencing where an event could be announced when focusing a name (e.g. "name" is speaking, where "is speaking" is inserted via the after atribute).

**Comment by jcsteh** [comment]
Is this failing in browse mode, focus mode or both?

In Firefox, the check box does get reported for me when I mouse click the label. I do intermittently see the state reported incorrectly, but the presence of ::before doesn't seem to make a difference.

My guess is that this has to do with the order of events being fired, since clicking the check box might focus it and then change its value or it might happen the other way around. Even so, NVDA should be reporting the change even if it does so afterward and it isn't. I'm not sure why.

**Comment by jcsteh** [comment]
I can only reproduce this in browse mode. In focus mode, this works as expected for me. That suggests the virtual buffer isn't being updated in time, maybe because the state change event is firing after the focus event.

**Comment by Adriani90** [comment]
> the presence of ::before doesn't seem to make a difference.

Yeah I think the before or after atribute has only an effect when you use focus mode and press space bar to change the state. This is not supported in browse mode or when using the mouse.

> I can only reproduce this in browse mode. In focus mode, this works as expected for me.

I see, I can reproduce this only sometimes though. And sometimes it works correctly both in focus and browse mode.

by the way, the only case where I can reproduce an inverted reporting of label and control state is when using the mouse and clicking on the label directly vs. using the mouse and clicking on the control which is at the left part of its label. This happens on both checkboxes.

**Comment by Adriani90** [comment]
Hmm this clicking on the label vs. clicking on the control itself behavior is quite weird. It seems when clicking on the control itself, the state is always reported correctly, double checked when pressing arrow keys in browse mode after changing the state with the mouse.
However, when clicking on the label, NVDA reports the previous state and not the current one. NVDA says for example
"normal checkbox not checked" although this was the state before I clicked on the label.

This seems quite severe.
@LeonarddeR is this maybe a regression introduced by #15518?

**Comment by Adriani90** [comment]
@franzitaz next time please include the expected reporting and the actual reporting as plain text and not only as screenshots. Here are a lot of blind developers and users like me for example and we cannot get the screenshots read by the screen reader. This caused some confusions on my side when testing.

**Comment by franzitaz** [comment]
> @franzitaz next time please include the expected reporting and the actual reporting as plain text and not only as screenshots. Here are a lot of blind developers and users like me for example and we cannot get the screenshots read by the screen reader. This caused some confusions on my side when testing.

Thank you very much for your feedback @Adriani90 !
I will update my report.

I am new on web accessibility, and I am studying to improve on this area.
Thank you again!
